core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -361,37 +454,35 @@ private void migrateFromV11() {
          Files.move(newFile.toPath(), file.toPath(), StandardCopyOption.REPLACE_EXISTING);
          //reopen the file
          channel = SecurityActions.openFileChannel(file);
-         //update file position
-         filePos = newFilePos;
          PERSISTENCE.persistedDataSuccessfulMigrated(cacheName);
       } catch (IOException e) {
          throw PERSISTENCE.persistedDataMigrationFailed(cacheName, e);
       }
    }
 
-   private ByteBuffer loadAndTransformLegacy(ByteBuffer buf, ByRef.Long offset, int length)
-         throws ClassNotFoundException, InterruptedException, IOException {
-      Object obj = loadLegacyObject(buf, offset, length);
-      return transformLegacyObject(obj);
-   }
-
-   private <T> ByteBuffer transformLegacyObject(T object) throws InterruptedException, IOException {
-      byte[] bytes = ctx.getPersistenceMarshaller().objectToByteBuffer(object);
-      return ByteBuffer.wrap(bytes);
+   private <T> T unmarshallObject(ByteBuffer buf, ByRef.Long offset, int length) throws ClassNotFoundException, IOException {
+      return unmarshallObject(buf, offset, length, false);
    }
 
    @SuppressWarnings("unchecked")
-   private <T> T loadLegacyObject(ByteBuffer buf, ByRef.Long offset, int length) throws ClassNotFoundException, IOException {
+   private <T> T unmarshallObject(ByteBuffer buf, ByRef.Long offset, int length, boolean legacyWrapperMissing) throws ClassNotFoundException, IOException {
       buf = readChannelUpdateOffset(buf, offset, length);
-      // Read using raw user marshaller without MarshallUserObject wrapping
-      byte[] bytes = buf.array();
-      Marshaller marshaller = ctx.getPersistenceMarshaller().getUserMarshaller();
-      try {
-         return (T) marshaller.objectFromByteBuffer(bytes, 0, length);
-      } catch (IllegalArgumentException e) {
-         // For internal cache key/values and custom metadata we need to use the persistence marshaller
-         return (T) ctx.getPersistenceMarshaller().objectFromByteBuffer(bytes, 0, length);
+      buf.flip();
+      byte[] bytes = new byte[buf.remaining()];
```

Comment:
Is there a reason we can't use the underlying byte[] anymore? As far as I can see the underlying byte[] isn't retained in the methods it is passed to.

Reply:
Tbh I wasn't sure how correct/safe the original approach was, hence the change. I can revert this though if it's not going to cause any issues.

Reply:
Yeah, it should be safe. The `objectFromByteBuffer` methods shouldn't retain a reference so we should be fine just using the underlying byte[] in the `ByteBuffer` and can avoid the extra byte[] allocation.

Reply:
Updated to just use `byte[] bytes = buf.array();`